Issue “does nothing other than call itself recursively” warning in more cases

cahilltr opened the pull request:
Fixes scala/bug#11614 by adding a recursive warning to functions with empty parentheses, covering a method with an apply.

som-snytt commented on src/compiler/scala/tools/nsc/typechecker/RefChecks.scala:
```diff
@@ -1145,11 +1145,12 @@ abstract class RefChecks extends Transform {
     // TODO: Move to abide rule. Also, this does not check that the def is final or not overridden, for example
     def checkInfiniteLoop(sym: Symbol, rhs: Tree): Unit =
-      if (!sym.isValueParameter && sym.paramss.isEmpty) {
+      if (!sym.isValueParameter && (sym.paramss.isEmpty || sym.paramss.forall(_.isEmpty))) {
```
`paramss.isEmpty` is now logically redundant. Probably it is not an optimization.

som-snytt commented on src/compiler/scala/tools/nsc/typechecker/RefChecks.scala:
```diff
@@ -1145,11 +1145,12 @@ abstract class RefChecks extends Transform {
+      if (!sym.isValueParameter && (sym.paramss.isEmpty || sym.paramss.forall(_.isEmpty))) {
         rhs match {
-          case t@(Ident(_) | Select(This(_), _)) if t hasSymbolWhich (_.accessedOrSelf == sym) =>
+          case t@(Ident(_) | Select(This(_), _) | Apply(Select(This(_), _), _)) if t hasSymbolWhich (_.accessedOrSelf == sym) =>
```
`t` is redundant because `rhs` is just a param. Note the outside parens were also redundant (as I learned recently).

som-snytt commented on test/files/run/issue-11614.scala:
```diff
@@ -0,0 +1,3 @@
+object Test extends App {
```
This should not be a `run` test. I would augment existing test in test/files/neg/t6276.scala

som-snytt commented on src/compiler/scala/tools/nsc/typechecker/RefChecks.scala:
```diff
@@ -1145,11 +1145,12 @@ abstract class RefChecks extends Transform {
     }
 
     // scala/bug#6276 warn for trivial recursion, such as `def foo = foo` or `val bar: X = bar`, which come up more frequently than you might think.
+    // scala/bug#11614 warns for trivial recursion around `def x(): X = x`, covering a method with an apply
```
This notation for the bug number is what should be used in the PR comment.

I see they already suggest on the next line that this should be an external lint, which today means scalafix. This change is helpful and benign, but it is still useful to add the scalafix and later delete this check.

cahilltr replied:
I updated the PR to remove the comment line.  That being said, I'm unsure of what the next steps are to handle this.  Should I create a PR in the [scalafix project](https://github.com/scalacenter/scalafix)?

som-snytt reviewed (changes requested):
Some tweaks.

som-snytt commented:
I removed the review at-notifications from the comment because I think the reviewers widget is the correct mechanism, at least for reviewers who are listed there. An explicit mention could summon an arbitrary outside reviewer, of course. Is that correct? review of comment by @SethTisue

SethTisue commented:
> I removed the review at-notifications from the comment because I think the reviewers widget is the correct mechanism, at least for reviewers who are listed there.

@cahilltr is just following what CONTRIBUTING.md currently recommends — a PR updating the advice there would be welcome

SethTisue commented:
@som-snytt re-review?

som-snytt reviewed (approved):
It does nothing other than call itself recursively, but what else is there to do during pandemic lockdown?

som-snytt commented:
Thanks! It would be nice if you squashed the two commits and also change the commit text from "fixes..." to "Add a recursive warning..." which are the descriptive words you use. They stopped linking the commit and issue using words in the title, although sometimes the commit text will mention the issue number. Recent PR 9168 shows that style.

Or maybe someone will merge this before you get a chance to do that. It's like when you go looking for your stuff and someone has done some tidying.

cahilltr committed:
Adds a recursive warning to functions with empty parentheses, covering a method with an apply.

Adds warning as per scala/bug#11614 by wrapping the `Select` case in an `Apply` and updating the if statement to check for an empty list or a list of empty lists.
